### DISCUSSION TITLE
Helper created by `Switch as X` integration disappears after HA Core upgrade

### DISCUSSION THREAD

**Discussion by gralin** [jpcornil-git/HA-mcp23017#13] [General]
How to reproduce the problem:

1. Change one of your Switch entities (MCP23017 output) to be shown as Light.

![setup](https://user-images.githubusercontent.com/3457048/203147441-ad8e27e7-4c7d-4a43-a11f-3aa85e2f6000.png)

2. Place the new entity (light helper) on your dashboard.

![normal](https://user-images.githubusercontent.com/3457048/203148312-29ab487d-61a8-4f54-9eb3-a52262903856.png)

3. Upgrade HA Core. During HA restart your entities won't be available.

![update](https://user-images.githubusercontent.com/3457048/203148394-bea9a032-9427-443a-9d11-5abcc544d707.PNG)

5. Check your entities after HA restart. In my case the new entity (light helper) is gone. 

![issue](https://user-images.githubusercontent.com/3457048/203148486-1e601617-09f0-460a-aa64-9f532041d909.PNG)

At first I thought it happens on each HA restart, but it actually hapens on each HA Core upgrade. Maybe during/after upgrade there is some cleanup mechanism and because MCP23017 is loading slowly, there is some race condition and the system detects the helper entity (light) as orphaned and deletes it? I didn't see any errors or warnings in logs, but maybe it's logged on lower level...

I didn't report this in HA repository yet, because it only happens for me in case of this integration. I tried reproducing the problem with with entities from other integration (Gree, a core integration) and the helper entities don't disappear.

**Comment by gralin** [comment]
I've been having this problem ever since I started [converting switches to lights](https://www.home-assistant.io/integrations/switch_as_x/), so it's not linked to any particular version of HA.

**Comment by jpcornil-git** [comment]
Hi @gralin,

I never used this [switch_as_x](https://www.home-assistant.io/integrations/switch_as_x/) integration (prefer to either adapt the name/icon directly in de dashboard or via [customization](https://www.home-assistant.io/docs/configuration/customizing-devices/), e.g. a homeassistant/customize section in my configuration.yaml)

You should may be add some logging before [L81 of /switch_as_x/__init__.py](https://github.com/home-assistant/core/blob/master/homeassistant/components/switch_as_x/__init__.py#L81) to trace this (likely place where switch_as_x config_entry is removed), e.g.:

```python
...
            _LOGGER.warning(
                f"[DEBUG] Removing config_entry {entry}",
            )
            device_registry.async_update_device(
                device_id, remove_config_entry_id=entry.entry_id
            )
...
```

**Comment by jpcornil-git** [comment]
Hi @gralin,
Have you been able to trace this further (race condition between switch_as_x initialisation and availability of this integration I guess, former relies on the latter) ?
Cheers,
jpc

**Comment by jpcornil-git** [comment]
For the records, fix #18 seems to address this issue as well
